### PR TITLE
replace implicit object with implicit val

### PR DESCRIPTION
**Pull request by fommil**
`implicit object` is evil

I'll keep pushing to this PR with changes towards #1823 ... my thinking on how typeclasses / implicit resolution works has matured a lot since writing stalactite.

### CODE REVIEW

File: core/src/main/scala/org/ensime/indexer/graph/GraphService.scala [hunk 1]
```diff
@@ -460,11 +460,10 @@ object GraphService {
   import org.ensime.util.stringymap.api._
   import org.ensime.util.stringymap.impl._
 
-  implicit object AccessSPrimitive extends SPrimitive[Access] {
+  implicit val access: SPrimitive[Access] = new SPrimitive[Access] {
     import org.objectweb.asm.Opcodes._
-    import SPrimitive.IntSPrimitive
 
-    def toValue(v: Access): java.lang.Integer =
+    def toValue(v: Access): AnyRef =
```
Comment (iravid) [inline core/src/main/scala/org/ensime/indexer/graph/GraphService.scala]: sigh

Comment (fommil) [reply core/src/main/scala/org/ensime/indexer/graph/GraphService.scala]: hehe, yeah, orient sucks.

File: json/src/test/scala/spray/json/BasicFormatsSpec.scala [hunk 2]
```diff
@@ -30,42 +30,18 @@ class BasicFormatsSpec extends WordSpec with DefaultJsonProtocol {
...
-    "convert a JsNull to a Float" in {
-      JsNull.convertTo[Float].isNaN shouldEqual Float.NaN.isNaN
-    }
   }
 
   "The DoubleJsonFormat" should {
     "convert a Double to a JsNumber" in {
       4.2.toJson shouldEqual JsNumber(4.2)
     }
-    "convert a Double.NaN to a JsNull" in {
```
Comment (vovapolu) [inline json/src/test/scala/spray/json/BasicFormatsSpec.scala]: Why did you remove these tests?

File: json/src/test/scala/spray/json/CollectionFormatsSpec.scala [hunk 3]
```diff
@@ -20,7 +18,9 @@ class CollectionFormatsSpec extends WordSpec with DefaultJsonProtocol {
     }
   }
 
+  /*
```
Comment (vovapolu) [inline json/src/test/scala/spray/json/CollectionFormatsSpec.scala]: Commented tests?

File: lsp/src/main/scala/org/ensime/lsp/rpc/RpcFormats.scala [hunk 4]
```diff
@@ -133,8 +141,8 @@ object RpcFormats {
 
   implicit val JsonRpcRequestMessageBatchFormat
     : RootJsonFormat[JsonRpcRequestMessageBatch] =
-    RpcConversions.JsonRpcRequestMessageBatchFormat
+    RpcConversions.request
   implicit val JsonRpcResponseMessageBatchFormat
     : RootJsonFormat[JsonRpcResponseMessageBatch] =
-    RpcConversions.JsonRpcResponseMessageBatchFormat
+    RpcConversions.response
```
Comment (vovapolu) [inline lsp/src/main/scala/org/ensime/lsp/rpc/RpcFormats.scala]: I think it's okay to have such long names for `*BatchFormat`, there're quite rare as I understand

File: lsp/src/test/scala/org/ensime/lsp/rpc/MessageCompanionsSpec.scala [hunk 5]
```diff
@@ -29,6 +28,9 @@ object MessageCompanionsSpec {
   }
 
   object Commands extends DefaultJsonProtocol with FamilyFormats {
+    override implicit def optionFormat[T: JsonFormat]: JsonFormat[Option[T]] =
```
Comment (vovapolu) [inline lsp/src/test/scala/org/ensime/lsp/rpc/MessageCompanionsSpec.scala]: It looks like a boilerplate :thinking:

File: lsp/src/test/scala/org/ensime/lsp/rpc/MessageCompanionsSpec.scala [hunk 6]
```diff
@@ -49,10 +51,13 @@ object MessageCompanionsSpec {
...
 
   object ResponsesFormats extends DefaultJsonProtocol with FamilyFormats {
+    override implicit def optionFormat[T: JsonFormat]: JsonFormat[Option[T]] =
+      super.optionFormat
+
     implicit val updateAccountCommand: RootJsonFormat[UpdateAccountCommand] =
       cachedImplicit
-    implicit val addTransactionCommand: RootJsonFormat[AddTransactionResponse] =
-      rootFormat(JsonUtils.wrapperFormat(AddTransactionResponse, _.id))
+    implicit val addTransactionCommand: JsonFormat[AddTransactionResponse] =
```
Comment (vovapolu) [inline lsp/src/test/scala/org/ensime/lsp/rpc/MessageCompanionsSpec.scala]: Not `RootJsonFormat`? Or we are going to remove it anyway?

File: json/src/main/scala/spray/json/CollectionFormats.scala [hunk 7]
```diff
@@ -2,42 +2,40 @@
 // License: http://www.gnu.org/licenses/lgpl-3.0.en.html
 package spray.json
 
-import scala.reflect.ClassTag
+import scala.collection.generic.CanBuildFrom
+import scala.language.higherKinds
 
 trait CollectionFormats {
 
+  implicit def cbf[T[_], A: JsonFormat](
```
Comment (vovapolu) [inline json/src/main/scala/spray/json/CollectionFormats.scala]: :+1: that's cool

**Comment by vovapolu** [comment]
`RootJsonFormat` is getting even messier...
I hope you will remove it soon :)

File: json/src/test/scala/spray/json/BasicFormatsSpec.scala (hunk 2 above)
Comment (fommil) [reply json/src/test/scala/spray/json/BasicFormatsSpec.scala]: We no longer support NaN since we go via BigDecimal.

File: json/src/test/scala/spray/json/CollectionFormatsSpec.scala (hunk 3 above)
Comment (fommil) [reply json/src/test/scala/spray/json/CollectionFormatsSpec.scala]: Was hoping to get it working but dropping array support seems best.

File: lsp/src/test/scala/org/ensime/lsp/rpc/MessageCompanionsSpec.scala (hunk 5 above)
Comment (fommil) [reply lsp/src/test/scala/org/ensime/lsp/rpc/MessageCompanionsSpec.scala]: This is the cost of the mixins. This problem goes away when moving to export-hook style (next on the list)

File: lsp/src/test/scala/org/ensime/lsp/rpc/MessageCompanionsSpec.scala (hunk 6 above)
Comment (fommil) [reply lsp/src/test/scala/org/ensime/lsp/rpc/MessageCompanionsSpec.scala]: It's not a root.

**Comment by fommil** [comment]
I might keep the root. Adding a new JsStringFormat would be good for the Map formatter... that's always been a weakness in json formatters.

File: lsp/src/test/scala/org/ensime/lsp/rpc/MessageCompanionsSpec.scala (hunk 6 above)
Comment (vovapolu) [reply lsp/src/test/scala/org/ensime/lsp/rpc/MessageCompanionsSpec.scala]: :thinking: why?

Comment (fommil) [reply lsp/src/test/scala/org/ensime/lsp/rpc/MessageCompanionsSpec.scala]: roots are only objects and arrays, but you're contramapping over something else here

**Comment by vovapolu** [comment]
Maybe let's add issue for this? I will come back to this after atom plugin (hope it won't take much time).

**Comment by fommil** [comment]
I'll eventually finish off this PR... just assume it doesn't exist.
